core/server/models/session.js

```diff
@@ -0,0 +1,81 @@
+const ghostBookshelf = require('./base');
+
+const Session = ghostBookshelf.Model.extend({
+    tableName: 'sessions',
+
+    parse(attrs) {
+        attrs.session_data = JSON.parse(attrs.session_data);
+        return attrs;
+    },
+
+    format(attrs) {
+        // CASE: format will be called when formatting all data for the DB
+        // including for SELECTs meaning that if we call findOne without
+        // a session_data property we'll get unintended JSON.stringify(undefined) calls
+        if (attrs.session_data) {
+            attrs.session_data = JSON.stringify(attrs.session_data);
+        }
+        return attrs;
+    },
+
+    user() {
+        return this.belongsTo('User');
+    }
+}, {
+
+    permittedOptions(methodName) {
+        const permittedOptions = ghostBookshelf.Model.permittedOptions.call(this, methodName);
+        if (methodName === 'upsert') {
+            return permittedOptions.concat('session_id');
+        }
+        if (methodName === 'destroy') {
+            return permittedOptions.concat('session_id');
+        }
+        return permittedOptions;
+    },
+
+    destroy(unfilteredOptions) {
+        if (unfilteredOptions.id) {
+            return ghostBookshelf.Model.destroy.call(this, unfilteredOptions);
+        }
+        const options = this.filterOptions(unfilteredOptions, 'destroy');
+
+        // Fetch the object before destroying it, so that the changed data is available to events
+        return this.forge({session_id: options.session_id})
+            .fetch(options)
+            .then((obj) => {
+                return obj.destroy(options);
+            });
+    },
+
+    upsert(data, unfilteredOptions) {
+        const options = this.filterOptions(unfilteredOptions, 'upsert');
+        const sessionId = options.session_id;
+        const sessionData = data.session_data;
+        const userId = sessionData.user_id;
+        return this.findOne({session_id: sessionId, user_id: userId})
+            .then((model) => {
+                if (model) {
+                    return this.edit({
+                        session_data: sessionData
+                    }, Object.assign(options, {
```

Comment:
Here and below we pass the filtered options to `edit`/`add` should we be passing the unfiltered ones and allowing those methods to filter themselves... not sure :thinking:

Reply:
in this case, it's weird yeah, because it's "upsert". We usually just pass the filtered options, because it's the same action.

Reply:
Hmm yeah - i'm not. I think this is fine for now, as we don't have any custom filtering for edit/add